Fix cluster default initialization #1

willyborankin opened the pull request:
### Description
Change the way how the security plugin initialize its configuration. 
The current solution uses poll to upload the default security configuration from each node in the cluster. As result it could lead to the races during initialization of the security index. 

To solve the problem with possible races the new solution uses cluster state and initializes the default configuration only on the management node and after that sends updated cluster state on all other nodes to load the configuration from the index.   

Initialization using cluster state makes things much more simpler and security admin shell script can be removed in the future.

the cluster state for the security plugin is this:
```json
"security": {
    "created": "2024-02-27T15:44:44.194028751Z",
    "configuration": {
      "actiongroups": {
        "hash": "219b41179866fc2419b736aff874a23f",
        "last_modified": null
      },
      "allowlist": {
        "hash": "522ce41bf5755eeb591d0161844830c0",
        "last_modified": null
      },
      "config": {
        "hash": "d97d664526b42ffa583dbb492742d3cc",
        "last_modified": null
      },
      "internalusers": {
        "hash": "f15fbfe131e0beb466313906d207f00b",
        "last_modified": null
      },
      "nodesdn": {
        "hash": "38e29064f142bfdf8632bc1a66cf8d00",
        "last_modified": null
      },
      "roles": {
        "hash": "db58ba06f214bfb53eee783378f5500c",
        "last_modified": null
      },
      "rolesmapping": {
        "hash": "f2b1cbdf73fa861f1321983cb0241964",
        "last_modified": null
      },
      "tenants": {
        "hash": "7942de6dc0f3bade63d8fc4b49d727d2",
        "last_modified": null
      },
      "whitelist": {
        "hash": "b73aadf21bca6a0de199121d982da460",
        "last_modified": null
      }
    }
  }
```

Next steps as a separate PRs:
  - Security configuration update on the manager node only and wait for confirmation from other nodes that configuration successfully updated on them.
  - Add migration from old version to new version with the cluster state support     
  -  Change existing documentation to reflect changes here: https://github.com/opensearch-project/security/blob/main/ARCHITECTURE.md#security-configuration
     

### Issues Resolved
- Resolves #3972 
- Partially resolve #3275

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

willyborankin commented:
Hi @cliu123, @cwperks, @DarshitChanpura, @davidlago, @peternied , @RyanL1997, @scrawfor99 and @reta,
Lets discuss the new way of the sec plugin configuration initialization. There are test failures I think I know where is the problem. But I think it makes sense to start discussions. Thank you.

willyborankin commented:
now all tests passed

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -299,7 +312,7 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
             );
             try {
-                Thread.sleep(500);
+                TimeUnit.MICROSECONDS.sleep(500);
```
Previously this was in milliseconds - this will increase the load on the cluster manager when its trying to get into a healthy state, why the change?

willyborankin replied:
good catch. my bad did not notice

peternied commented on src/main/java/org/opensearch/security/state/SecurityMetadata.java:
```diff
@@ -0,0 +1,81 @@
+
+    public final static String SECURITY_CONFIGURATION_APPLIED_FIELD_NAME = "security_configuration_applied";
+
+    private final boolean securityConfigurationApplied;
```
! This is a great, can I ask you to make a tweak that will add a feature [1] we've been looking for?

Instead of a boolean if a identifier is associated associated with it, we can make sure the security configuration is the same version on all nodes.  

Note; there is more work associated with getting the security config refresh process connected to this, but I'd be happy to do that wiring in a separate follow up pull request.

- https://github.com/opensearch-project/security/issues/3275

willyborankin replied:
WDYT to add a property `versions` so cluster state will looks like:
```json
"security": {
    "initia_security_configuration_applied": true,
    "versions": {
         "actions_group": 1,
         "roles": 1,
         ....
     }
}
```
so if a new node joins, to the cluster, we will see versions for all `CTypes` the index contains. AFAIU for the current logic, there is no way to get this info. As a result, it is possible to upgrade the `CTypes` configuration as soon as the cluster switches to the new manager with the new configuration?

peternied replied:
Sure - that works too

peternied commented on src/main/java/org/opensearch/security/state/SecurityClusterNonManagerListener.java:
```diff
@@ -0,0 +1,50 @@
+        final var maxNodeVersion = clusterChangedEvent.state().nodes().getMaxNodeVersion();
+        if (clusterChangedEvent.state().custom(SecurityMetadata.TYPE, SecurityMetadata.DEFAULT).isSecurityApplied()) {
+            configurationRepository.loadConfiguration("Default security configuration applied successfully", CType.values());
+        } else if (!minNodeVersion.equals(maxNodeVersion)) {
```
Lets add a BWC test that can exercises this scenario to confirm everything works as expected.

willyborankin replied:
Sure

peternied commented on src/main/java/org/opensearch/security/state/SecurityClusterNonManagerListener.java:
```diff
@@ -0,0 +1,50 @@
+            // exists in the cluster.
+            // when a new cluster manager will be selected with the new version of the sec plugin cluster state will be
+            // updated but it won't reload the configuration
+            configurationRepository.loadConfiguration(
```
Lets add a log info message if we are in this state

peternied commented on src/main/java/org/opensearch/security/state/SecurityClusterNonManagerListener.java:
```diff
@@ -0,0 +1,50 @@
+        }
+        final var minNodeVersion = clusterChangedEvent.state().nodes().getMinNodeVersion();
+        final var maxNodeVersion = clusterChangedEvent.state().nodes().getMaxNodeVersion();
+        if (clusterChangedEvent.state().custom(SecurityMetadata.TYPE, SecurityMetadata.DEFAULT).isSecurityApplied()) {
```
Will this be triggered on any cluster change event, or only changes to this key?  Seems like there should be some filtering before we reload the configuration.

willyborankin replied:
it triggers on each cluster state change. It is how OS works with the state.

peternied commented on src/main/java/org/opensearch/security/state/SecurityClusterManagerListener.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public void onClusterManager() {
+        if (!clusterService.state().custom(SecurityMetadata.TYPE, SecurityMetadata.DEFAULT).isSecurityApplied()) {
```
For readability
```suggestion
        final boolean securityIsUninitialized = !clusterService.state().custom(SecurityMetadata.TYPE, SecurityMetadata.DEFAULT).isSecurityApplied();
        if (securityIsUnitialized) {
```

peternied commented on src/main/java/org/opensearch/security/state/SecurityClusterManagerListener.java:
```diff
@@ -0,0 +1,75 @@
+    public void onClusterManager() {
+        if (!clusterService.state().custom(SecurityMetadata.TYPE, SecurityMetadata.DEFAULT).isSecurityApplied()) {
+            threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(() -> {
+                configurationRepository.initDefaultConfiguration();
```
What do you think about adding some way to validate this was completed successful. what about error cases?

peternied commented on src/main/java/org/opensearch/security/state/SecurityClusterManagerListener.java:
```diff
@@ -0,0 +1,75 @@
+
+public class SecurityClusterManagerListener implements LocalNodeClusterManagerListener {
+
+    private static final Logger LOGGER = LogManager.getLogger(ConfigurationRepository.class);
```
Wrong logger?

peternied commented on src/main/java/org/opensearch/security/securityconf/impl/CType.java:
```diff
@@ -50,21 +51,24 @@
-    WHITELIST(toMap(1, WhitelistingSettings.class, 2, WhitelistingSettings.class)),
-    ALLOWLIST(toMap(1, AllowlistingSettings.class, 2, AllowlistingSettings.class)),
-    AUDIT(toMap(1, AuditConfig.class, 2, AuditConfig.class));
+    ACTIONGROUPS(toMap(0, List.class, 1, ActionGroupsV6.class, 2, ActionGroupsV7.class), "action_groups.yml"),
```
I like this - a same it wasn't in alpha order to begin with

willyborankin replied:
:-)

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +325,58 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+                    ConfigHelper.uploadFile(client, CType.ROLESMAPPING, configDir, securityIndex, DEFAULT_CONFIG_VERSION);
+                    ConfigHelper.uploadFile(client, CType.INTERNALUSERS, configDir, securityIndex, DEFAULT_CONFIG_VERSION);
+                    ConfigHelper.uploadFile(client, CType.ACTIONGROUPS, configDir, securityIndex, DEFAULT_CONFIG_VERSION);
+                    // FIXME remove it as soon as we will read of version 1 config
```
Is there a follow up issue for this comment?

willyborankin replied:
AFAIR @cwperks already created an issue for it. I think I will update it.

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +325,58 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+                LOGGER.error("{} does not exist", CType.CONFIG.configFile(configDir).toAbsolutePath());
+            }
+        } catch (Exception e) {
+            LOGGER.error("Cannot apply default config (this is maybe not an error!)", e);
```
`(this is maybe not an error!)` Can we remove this, or handle non-error cases better?

willyborankin replied:
sure I will

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +325,58 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+        AccessController.doPrivileged((PrivilegedAction<Boolean>) () -> {
+            try {
+                if (!clusterService.state().metadata().hasIndex(securityIndex)) {
+                    initalizeConfigTask.complete(null);
```
Shouldn't this be triggered after line 336?

willyborankin replied:
It should be

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +325,58 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+                if (!clusterService.state().metadata().hasIndex(securityIndex)) {
+                    initalizeConfigTask.complete(null);
+                    LOGGER.info("Creating index {} and default configs if they are absent", securityIndex);
+                    createSecurityIndexIfAbsent();
```
Can this replace the condition at 331?

willyborankin replied:
Yes.

peternied reviewed (commented):
Nice work @willyborankin!

Overall comment, it looks like much of code you've added isn't getting registering code coverage hits - something might be wrong with the test execution, can you look into that?

willyborankin commented:
Hi @peternied, thanks for your feedback. I will reorganize the code a bit. The way I implemented the index initialization is a bit wrong. The initialization needs to be done by the generic executor and changing of the cluster state via management executor. The management executor has a 20-second timeout, so it could lead to problems if the index initialization takes more than 20 seconds.

willyborankin commented:
Hi @peternied, the latest version which includes:
 - Adding of configuration now use bulk update  
 - Any node fails if it can't read/write configuration 
- Addressed all comments you mentioned.

I think we can extend tests for the functionality I added. Lets discuss additional tests.

willyborankin commented on src/main/java/org/opensearch/security/state/SecurityClusterStateListeners.java:
```diff
@@ -0,0 +1,128 @@
+                            + "Default configuration will be re-init from scratch on the next selected management node."
+                    );
+                    initDefaultConfigFuture.cancel(true);
+                    initDefaultConfigFuture = null;
```
not sure that this is the right way to stop future

peternied replied:
I'm not clear what the intention behind initDefaultConfigFuture, can you help me understand how you'd like the flow of control to work?

willyborankin replied:
Ahhh I observed it when I re-ran integration tests 100 times. Sometimes we have problems with ports as result the manager node startes the initialization and after that switches to another node which could be the test framework problem we use. I haven't seen such problems with in the real env.  So I will remove it so far and in case there is such problem I will add such logic.

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +358,75 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+
+    // visible for testing
+    protected Future<Void> reloadConfiguration(final SecurityMetadata securityMetadata) {
+        if (!initalizeConfigTask.isDone() && !configReloadInProcess.get()) {
+            configReloadInProcess.getAndSet(true);
```
This isn't thread safe, see https://stackoverflow.com/questions/4501223/when-do-i-need-to-use-atomicboolean-in-java

peternied replied:
Compare & set should happen in the same function call, e.g

```
if (configReloadInProcess.compareAndSet(false, true)) {
    // initialize ...
}
```

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +358,75 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+            configReloadInProcess.getAndSet(true);
+            return threadPool.generic().submit(() -> {
+                securityIndexHandler.loadConfiguration(securityMetadata.configuration(), ActionListener.wrap(cTypeConfigs -> {
+                    configCache.putAll(cTypeConfigs);
```
Seems like this is duplication logic from `reloadConfiguration0`, can we reduce this or make the shared usages cleaner?

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +358,75 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
 
+    // viable for testing
+    protected Future<Void> initDefaultConfiguration() {
+        return threadPool.generic().submit(() -> {
```
Can `securityIndexHandler.createIndex(...)` or `uploadSecurityConfiguration()` be called without a thread, they already use listeners for async actions?

willyborankin replied:
Unfortunately no :-( all operations withing event listener must be called in the thread pool, otherwise it fails.

willyborankin replied:
But I will change it. In our case we need to trigger initialization when we already have a manager node.

peternied commented on src/main/java/org/opensearch/security/support/SecurityIndexHandler.java:
```diff
@@ -0,0 +1,247 @@
+    ) {
+        try (final ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+            client.threadPool().getThreadContext().putHeader(ConfigConstants.OPENDISTRO_SECURITY_CONF_REQUEST_HEADER, "true");
+            final var refresh = configuration.stream().map(c -> c.lastModified().isPresent()).findFirst().orElse(false);
```
Or maybe if we ever made a request and got back a version that was different than the expected last modified, seems like we should re-issue the configuration reload

willyborankin replied:
Yes ... overdid it :-)

peternied commented on src/main/java/org/opensearch/security/support/ConfigHelper.java:
```diff
@@ -57,6 +57,7 @@
 
 import static org.opensearch.core.xcontent.DeprecationHandler.THROW_UNSUPPORTED_OPERATION;
 
+@Deprecated
```
Any reason not to delete?

willyborankin replied:
Yes so far I split up logic with the cluster state update and existing one. So in case of any problem we can switch back to old the logic.

willyborankin commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -119,8 +142,30 @@ private ConfigurationRepository(
+    public void applyClusterState(final ClusterChangedEvent event) {
+        boolean wasClusterManager = event.previousState().nodes().isLocalNodeElectedClusterManager();
+        boolean isClusterManager = event.localNodeClusterManager();
+        if (!wasClusterManager && isClusterManager) { // on manager node only!!!
```
Good point I changed it a bit. It will be triggered when node is a manager. Regarding docs I added the 3rd step to change documentation.

willyborankin commented on src/test/java/org/opensearch/security/support/SecurityIndexHandlerTest.java:
```diff
@@ -0,0 +1,418 @@
+    @SuppressWarnings("unchecked")
+    public void createIndex_shouldReturnSecurityExceptionIfItCannotCreateIndex() {
+
+        final var listener = mock(ActionListener.class);
```
Good point! AI always forget about `spy` :-)

willyborankin commented on src/main/java/org/opensearch/security/support/SecurityIndexHandler.java:
```diff
@@ -0,0 +1,247 @@
+    private String configTypeSource(final InputStream inputStream) throws IOException {
+        final var jsonContent = XContentType.JSON.xContent();
+        try (final var parser = jsonContent.createParser(NamedXContentRegistry.EMPTY, THROW_UNSUPPORTED_OPERATION, inputStream)) {
+            parser.nextToken();
+            parser.nextToken();
+            parser.nextToken();
```
I was thinking to replace it with Jackson, but I think we can do it as a follower PR. wdyt?

willyborankin commented on src/main/java/org/opensearch/security/support/SecurityIndexHandler.java:
```diff
@@ -0,0 +1,247 @@
+                                "WHITELIST configuration type is deprecated and will be replaced with ALLOWLIST in the next major version"
+                            );
+                        }
+                        final var yamlContent = yamlContentFor(cType, configDir);
```
Jackson does not support mix of parsers. It works only with one of them. In this case we need convert file content to YAML and after that to JSON. I'm planning to looks at it a bit later maybe it is possible to improve it.

willyborankin commented on src/test/java/org/opensearch/security/configuration/ConfigurationRepositoryTest.java:
```diff
@@ -111,4 +151,165 @@ public void getConfiguration_withInvalidConfigurationShouldReturnNewEmptyConfigu
+            configRepository.reloadConfiguration(new SecurityMetadata(Instant.now(), Set.of(new SecurityConfig(CType.CONFIG, "aaa", null))))
+        );
+
+        verify(securityIndexHandler, never()).loadConfiguration(any(), any());
```
Addressed

DarshitChanpura replied:
`verifyNoMoreInteractions(...)` still seems to be part of each individual test. Maybe the code-commit push was missed?

willyborankin replied:
I added verifyNoMoreInteractions for all tests except: 

- testClusterChanged_shouldInitSecurityIndexIfNoSecurityData
- testClusterChanged_shouldExecuteInitialization
- testClusterChanged_shouldNotExecuteInitialization

since all of them do `doCallRealMethod`

willyborankin commented:
All comments I addressed :-)

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -119,8 +141,33 @@ private ConfigurationRepository(
+            }
+        }
+        if (securityMetadata != null) {
+            executeConfigurationInitialization(event.state().custom(SecurityMetadata.TYPE));
```
```suggestion
            executeConfigurationInitialization(securityMetadata);
```

willyborankin replied:
fixed

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +360,71 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
         }
     }
 
+    // viable for testing
```
Nit
```suggestion
```

willyborankin replied:
Oops

willyborankin replied:
fixed

peternied commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -312,6 +360,71 @@ private void waitForSecurityIndexToBeAtLeastYellow() {
+                });
+            }
+        }
+        return null;
```
I'd recommend return a complete*d* future so callers don't have to worry about nulls
```suggestion
        return CompletableFuture.completedFuture(null);
```

willyborankin commented:
There is one failure which is related to https://github.com/opensearch-project/security/issues/4077. The problem is on OS side. As soon as it will be fixed PR should be green

peternied reviewed (commented):
Looking good

peternied reviewed (dismissed):
Is there a PR to resolve that issue [1] that this is blocked waiting on from OpenSearch? 

- [1] https://github.com/opensearch-project/security/issues/4077

willyborankin commented:
> Is there a PR to resolve that issue [1] that this is blocked waiting on from OpenSearch?
> 
>     * [1] [[BUG] Defaut tests which extend OpenSearchTestCase don't work #4077](https://github.com/opensearch-project/security/issues/4077)

So far. I did not have tie to test the fix which @reta suggested. I'm planning to do it tomorrow morning and based of results open PR in OS.

peternied commented:
@willyborankin While we wait for the fix, lets disable the test case to get this change in.  I've just pushed a commit to do this.

willyborankin commented:
> @willyborankin While we wait for the fix, lets disable the test case to get this change in. I've just pushed a commit to do this.

Sounds reasonable

willyborankin commented:
@peternied codecov is complaining about `SecutiyConfig` which is expected since the test was disabled

peternied commented:
@cwperks or @reta Mind giving this PR a look?

cwperks commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -119,8 +141,33 @@ private ConfigurationRepository(
+            }
+        }
+        if (securityMetadata != null) {
+            executeConfigurationInitialization(securityMetadata);
```
This looks much simpler. Nice!

Just to make sure I'm understanding this correctly, one of the primary purposes of `executeConfigurationInitialization` is to notify all of the nodes to populate their caches? 

On first bootstrap, `initSecurityIndex` is called since securityMetadata is null and it triggers [clusterService.submitStateUpdateTask](https://github.com/opensearch-project/security/pull/4002/files#diff-3dfaf08b5fd593ac73bb4955f7f579919ac870b5ed060636a2320e3ef85ef935R379-R387) which will again hit this block, but the second time it hits this block the securityMetadata will be non-null so it triggers this call?

willyborankin replied:
Exactly. As soon as cluster has a manager. The manager node triggers index initialization and if all is good it changes the cluster state so all nodes including manages will reload the cache. The cluster state update needs to be improved a bit but it wont change the current logic we have right now.

willyborankin replied:
done

DarshitChanpura commented on src/test/java/org/opensearch/security/support/SecurityIndexHandlerTest.java:
```diff
@@ -0,0 +1,384 @@
+/*
+ * Copyright 2015-2018 _floragunn_ GmbH
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
This license header should be removed.

willyborankin replied:
fixed

DarshitChanpura commented on src/test/java/org/opensearch/security/support/SecurityIndexHandlerTest.java:
```diff
@@ -0,0 +1,384 @@
+    }
+
+    @Test
+    public void uploadDefaultConfiguration_shouldFailIfBulkHasFailures() throws IOException {
```
is the bulk transaction atomic? or is partial success allowed ?

willyborankin replied:
OS does not support transactions. If bulk insertion fails the cluster state never updates state as result no cache reload on nodes.

DarshitChanpura replied:
So no half updates would be allowed. Meaning `roles` upload didn't fail but `rolesmapping` did, then both with fail? essentially, all or nothing?

willyborankin replied:
Half update is not allowed. It depends what the end user wanna to upload. The current solution is to check config.yml file only while the new one checks the list of important files the importance of the file is in the `CType` enum aka `allowEpmtyFile`

willyborankin commented on src/test/java/org/opensearch/security/configuration/ConfigurationRepositoryTest.java:
```diff
@@ -111,4 +164,143 @@ public void getConfiguration_withInvalidConfigurationShouldReturnNewEmptyConfigu
+    }
+
+    @Test
+    public void clusterChanged_shouldNotInitSecurityIndexIfNoSecurityDataExists() {
```
Agree the test is useless and confusing :-)

willyborankin commented on src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java:
```diff
@@ -231,42 +278,43 @@ private void initalizeClusterConfiguration(final boolean installDefaultConfig) {
 
+    private void setupAuditConfigurationIfAny(final boolean auditConfigDocPresent) {
+        final Set<String> deprecatedAuditKeysInSettings = AuditConfig.getDeprecatedKeys(settings);
+        if (!deprecatedAuditKeysInSettings.isEmpty()) {
```
Not sure. I think we should clean all deprivations in 3.x

DarshitChanpura replied:
created an issue: https://github.com/opensearch-project/security/issues/4117

DarshitChanpura commented on src/main/java/org/opensearch/security/support/ConfigReader.java:
```diff
@@ -0,0 +1,92 @@
+import static org.opensearch.core.xcontent.DeprecationHandler.THROW_UNSUPPORTED_OPERATION;
+import static org.opensearch.security.configuration.ConfigurationRepository.DEFAULT_CONFIG_VERSION;
+
+public final class ConfigReader {
```
could you add brief java-doc for this class?

DarshitChanpura commented on src/main/java/org/opensearch/security/support/SecurityIndexHandler.java:
```diff
@@ -0,0 +1,233 @@
+import static org.opensearch.security.support.ConfigReader.emptyJsonConfigFor;
+import static org.opensearch.security.support.ConfigReader.yamlContentFor;
+
+public class SecurityIndexHandler {
```
There are bunch of lines in this code not covered by code-cov, could you please look into that? Probly needs a unit test file.

willyborankin replied:
this is strange since `SecurityIndexHandlerTest` tests all methods. Do not know why we have such statistics :-(

peternied replied:
I just ran the coverage numbers locally, I see the same thing on the jacoco report

`./gradlew test --tests org.opensearch.security.support.SecurityIndexHandlerTest jacocoTestReport`  Pull up the local report with `build/reports/jacoco/test/html/org.opensearch.security.support/SecurityIndexHandler.java.html`

Big area that isn't covered IMO is buildDynamicConfiguration

cwperks reviewed (approved):
Thank you @willyborankin , I don't have any further comments on this PR. This PR looks well tested and the new functionality is behind a feature flag that makes it contained.

I do have a few questions about backward compatibility that could translate into tests using the bwc test framework:

- Is it possible to migrate from bootstrap w/o using cluster state -> bootstrap w/ cluster state with a rolling restart?
- Same, as above but with full cluster reboot

willyborankin commented:
> Thank you @willyborankin , I don't have any further comments on this PR. This PR looks well tested and the new functionality is behind a feature flag that makes it contained.
> 
> I do have a few questions about backward compatibility that could translate into tests using the bwc test framework:
> 
>     * Is it possible to migrate from bootstrap w/o using cluster state -> bootstrap w/ cluster state with a rolling restart?
So far no. The upgrade of existing clusters and such corner case I will prepare as a final step :-). The current solution works only for new clusters. if you disable cluster state all listeners won't be added to the `ClusterService` so you can use existing solution.  
>    * Same, as above but with full cluster reboot
Same as above.

In any case I think we need to introduce the full cluster state feature in a preview mode. wdyt?

peternied reviewed (commented):
It looks like there are some style / test failures.  Happy to re-approve when green.

willyborankin commented:
> It looks like there are some style / test failures. Happy to re-approve when green.

done.

willyborankin commented:
@peternied Looks like tests failures are note related. the new cluster permission appears:
```
cluster:monitor/term
```

willyborankin commented:
@peternied now all tests passed

willyborankin commented:
@peternied I think coverage now around 99 or 98 %

willyborankin commented:
@peternied the final score is  87.6 which is good I think

DarshitChanpura reviewed (approved):
Changes look good to me! Thank you for this change and adding a comprehensive test suite @willyborankin!

nibix commented:
@willyborankin quick question on this:

> Next steps as a separate PRs:
> Security configuration update on the manager node only and wait for confirmation from other nodes that configuration successfully updated on them.

Has this been implemented yet? 

I am asking because I need to refactor things a bit for https://github.com/opensearch-project/security/issues/5464#issuecomment-3068155867
